python-libraries/data-platform-catalogue/data_platform_catalogue/client.py

```diff
@@ -201,47 +310,277 @@ def get_user_id(self, user_email: str):
 
-    def _generate_duration(self, duration_in_days: int | None):
-        if duration_in_days is None:
-            return None
+            return data_product_urn
         else:
-            return Duration.parse_obj(f"P{duration_in_days}D")
+            raise ReferencedEntityMissing("Data Product must belong to a Domain")
+
+    def upsert_table(  # type: ignore[override]
+        self,
+        metadata: TableMetadata,
+        data_product_metadata: DataProductMetadata | None = None,
+        platform: str = "glue",
+        version: int = 1,
+    ):
+        """
+        Define a table (a 'dataset' in DataHub parlance), a 'collection of data'
+        (https://datahubproject.io/docs/metadata-modeling/metadata-model#the-core-entities).
+
+        There can be many tables per Data Product.
+
+        Columns are expected to be a list of dicts in the format
+            {"name": "column1", "type": "string", "description": "just an example"}
+
+        This method creates a schemaMetadata aspect object from the metadata object
+        (https://datahubproject.io/docs/generated/metamodel/entities/dataset/#schemametadata)
+        together with generating a unique reference name (URN) for the dataset used by DataHub.
+        These are then emitted to the rest.li api as a dataset creation/update event proposal.
+
+        If tags are present in the metadata object, a second request is made to update the dataset
+        with these tags as a separate aspect.
+
+        Args:
+            metadata (TableMetadata): metadata object.
+            platform (str, optional): DataHub data platform type. Defaults to "glue".
+            version (int, optional): Defaults to 1.
+
+        Returns:
+            dataset_urn: the dataset URN
+        """
+        dataset_schema_properties = SchemaMetadataClass(
+            schemaName=metadata.name,
+            platform=make_data_platform_urn(platform=platform),
+            version=version,
+            hash="",
+            platformSchema=OtherSchemaClass(rawSchema="__insert raw schema here__"),
+            fields=[
+                SchemaFieldClass(
+                    fieldPath=f"{column['name']}",
+                    type=SchemaFieldDataTypeClass(
+                        type=DATAHUB_DATA_TYPE_MAPPING[column["type"]]
+                    ),
+                    nativeDataType=column["type"],
+                    description=column["description"],
+                )
+                for column in metadata.column_details
+            ],
+        )
+
+        dataset_urn = mce_builder.make_dataset_urn(
+            platform=platform, name=f"{metadata.name}", env="PROD"
+        )
+
+        metadata_event = MetadataChangeProposalWrapper(
+            entityType="dataset",
+            changeType=ChangeTypeClass.UPSERT,
+            entityUrn=dataset_urn,
+            aspect=dataset_schema_properties,
+        )
+        self.graph.emit(metadata_event)
```

Comment:
⚠️ From integration testing this, it looks like the `SchemaPropertiesClass` we pass here is not enough for the dataset to show up in the UI. I *think* we need a DatasetPropertiesClass aspect as well, i.e. this snippet

```python
dataset_properties = DatasetPropertiesClass(
    description="Example table for testing",
    customProperties={"dpia_required": "yes"},
)

# Construct a MetadataChangeProposalWrapper object.
metadata_event = MetadataChangeProposalWrapper(
    entityUrn=builder.make_dataset_urn(
        "athena", "athena.catalog.hmpps_use_of_force_v2.statements"
    ),
    aspect=dataset_properties,
)
```